content/blog/nats-operator-and-service-accounts.md

```diff
@@ -0,0 +1,337 @@
+relies on the `TokenRequest` and `PodShareProcessNamespace` alpha features
+that were made available in the Kubernetes v1.10 release.
+
+The [TokenRequest API](https://github.com/kubernetes/community/blob/4a41674f6fb6fa69b3dfaf15267d908e909685d7/contributors/design-proposals/auth/bound-service-account-tokens.md) feature which allows issuing new tokens related
+to a `ServiceAccount` but with a different audience as the one used for
+the API server.  _Customizing the audience_ is very important in this
+case, since using the bare original token from `ServiceAccount` can
+make us run into security issues since this means that any container
+with the token would have the same permissions to get access to the
+Kubernetes API.
+
+```go
+// Issue token with audience set for the NATS cluster in this namespace only,
+// this will prevent the token from being usable against the API Server.
+ar := &authenticationv1.TokenRequest{
+	Spec: authenticationv1.TokenRequestSpec{
+		Audiences: []string{fmt.Sprintf("nats://%s.%s.svc", clusterName, ns)},
+        // ...
+	},
+}
+tr, err := kubecli.ServiceAccounts(ns).CreateToken(sa.Name, ar)
+if err != nil {
+	return err
+}
+```
+
+This token is then stored in a secret which can be mounted by a Pod
+(relatedly, there is a new feature in Kubernetes v1.11 called [Service
+Account Volume Projections](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-volume-projection) which would make it possible to not having
+to rely on a secret this way but more on that below).  Then the
+configuration of the NATS Server is updated including the newly mapped
+service account + nats service role.
+
+To be able to update the configuration of a cluster with zero
+downtime, assuming that the [PodShareProcessNamespace](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) feature flag has
+been enabled, the NATS Operator will enable the `ShareProcessNamespace`
+flag for each one of the NATS Pods.
+
+```go
+// Enable PID namespace sharing and attach sidecar that
+// reloads the server whenever the config file is updated.
+if cs.Pod != nil && cs.Pod.EnableConfigReload {
+	pod.Spec.ShareProcessNamespace = &[]bool{true}[0]
+	// ...
+}
+```
+
+Usually when running the NATS Server in Docker, it would run as PID 1
+inside of the container but in this case since PID namespace sharing
+has been enabled it is logging `6` instead below:
+
+```
+kubectl -n nats-io logs example-nats-1 nats | head
+[6] [INF] Starting nats-server version 1.2.0
+[6] [INF] Git commit [6608e9a]
+[6] [INF] Starting http monitor on 0.0.0.0:8222
+[6] [INF] Listening for client connections on 0.0.0.0:4222
+[6] [INF] Server is ready
+[6] [INF] Listening for route connections on 0.0.0.0:6222
+```
+
+Using `kubectl` we can confirm see there are `2/2` containers running
+for each one of the pods as well (nats server + reloader sidecar).
+
```

Comment:
sing `kubectl` we can confirm there are `2/2` containers running
+for each one of the pods as well (nats server + reloader sidecar).